scenario/state.py

```diff
@@ -1423,12 +1423,28 @@ def deferred(self, handler: Callable, event_id: int = 1) -> DeferredEvent:
         )
 
 
+_next_action_id_counter = 1
+
+
+def next_action_id(update=True):
+    global _next_action_id_counter
+    cur = _next_action_id_counter
+    if update:
+        _next_action_id_counter += 1
+    # Juju currently uses numbers for the ID, but in the past used UUIDs, so
+    # we need these to be strings.
+    return str(cur)
+
+
 @dataclasses.dataclass(frozen=True)
 class Action(_DCBase):
     name: str
 
     params: Dict[str, "AnyJson"] = dataclasses.field(default_factory=dict)
 
+    id: str = dataclasses.field(default_factory=next_action_id)
+    """Juju action ID. Every action invocation gets a new unique one."""
```

Comment:
I'd add a note like for relation.id, hinting at the fact that typically you won't need to worry about it unless you're testing something very specific. (I guess I could be a bit more explicit in relation.id as well)